File: python/samples/02-agents/structured_output.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio

from agent_framework import AgentResponse
from agent_framework.openai import OpenAIResponsesClient
from pydantic import BaseModel

"""
Structured Output from an Agent

Demonstrates how to get typed, structured responses from an agent using Pydantic models.
The agent returns data that is automatically parsed into your model class, enabling
type-safe access to response fields.

Covers both non-streaming and streaming modes with response_format.

For more on structured output:
- Deeper sample: getting_started/agents/openai/openai_responses_client_with_structured_output.py
- Azure AI response format: getting_started/agents/azure_ai/azure_ai_with_response_format.py
- Docs: https://learn.microsoft.com/agent-framework/concepts/structured-output
"""


# <output_model>
class CityInfo(BaseModel):
    """Structured output model for city information."""

    city: str
    country: str
    description: str
# </output_model>


async def non_streaming_example() -> None:
    """Get structured output from a non-streaming agent call."""
    print("=== Non-streaming Structured Output ===")

    # <create_agent>
    agent = OpenAIResponsesClient().as_agent(
        name="CityAgent",
        instructions="You are a helpful agent that describes cities in a structured format.",
    )
    # </create_agent>

    query = "Tell me about Paris, France"
    print(f"User: {query}")

    # <structured_response>
    result = await agent.run(query, options={"response_format": CityInfo})

    if structured_data := result.value:
        print(f"City: {structured_data.city}")
        print(f"Country: {structured_data.country}")
        print(f"Description: {structured_data.description}")
    else:
        print(f"Could not parse response: {result.text}")
    # </structured_response>


async def streaming_example() -> None:
    """Get structured output from a streaming agent call."""
    print("\n=== Streaming Structured Output ===")

    agent = OpenAIResponsesClient().as_agent(
        name="CityAgent",
        instructions="You are a helpful agent that describes cities in a structured format.",
    )

    query = "Tell me about Tokyo, Japan"
    print(f"User: {query}")

    # <streaming_structured>
    result = await AgentResponse.from_update_generator(
        agent.run(query, stream=True, options={"response_format": CityInfo}),
        output_format_type=CityInfo,
    )

    if structured_data := result.value:
        print(f"City: {structured_data.city}")
        print(f"Country: {structured_data.country}")
        print(f"Description: {structured_data.description}")
    else:
        print(f"Could not parse response: {result.text}")
    # </streaming_structured>


async def main() -> None:
    await non_streaming_example()
    await streaming_example()


if __name__ == "__main__":
    asyncio.run(main())

File: dotnet/samples/02-agents/StructuredOutput.cs
// Copyright (c) Microsoft. All rights reserved.

// Structured Output
// Configure an agent to produce structured (typed) JSON output.
// Demonstrates both RunAsync<T> and streaming with deserialization.
//
// For docs: https://learn.microsoft.com/agent-framework/agents/structured-output

using System.ComponentModel;
using System.Text.Json;
using System.Text.Json.Serialization;
using Azure.AI.Projects;
using Azure.Identity;
using Microsoft.Agents.AI;
using SampleApp;

#pragma warning disable CA5399

string endpoint = Environment.GetEnvironmentVariable("AZURE_FOUNDRY_PROJECT_ENDPOINT")
    ?? throw new InvalidOperationException("AZURE_FOUNDRY_PROJECT_ENDPOINT is not set.");
string deploymentName = Environment.GetEnvironmentVariable("AZURE_FOUNDRY_PROJECT_DEPLOYMENT_NAME") ?? "gpt-4o-mini";

AIProjectClient aiProjectClient = new(new Uri(endpoint), new AzureCliCredential());

// <structured_output>
ChatClientAgent agent = await aiProjectClient.CreateAIAgentAsync(
    model: deploymentName,
    new ChatClientAgentOptions()
    {
        Name = "StructuredOutputAssistant",
        ChatOptions = new()
        {
            Instructions = "You are a helpful assistant that extracts structured information about people.",
            ResponseFormat = Microsoft.Extensions.AI.ChatResponseFormat.ForJsonSchema<PersonInfo>()
        }
    });

AgentResponse<PersonInfo> response = await agent.RunAsync<PersonInfo>(
    "Please provide information about John Smith, who is a 35-year-old software engineer.");

Console.WriteLine($"Name: {response.Result.Name}");
Console.WriteLine($"Age: {response.Result.Age}");
Console.WriteLine($"Occupation: {response.Result.Occupation}");
// </structured_output>

// <structured_output_streaming>
IAsyncEnumerable<AgentResponseUpdate> updates = agent.RunStreamingAsync(
    "Please provide information about Jane Doe, who is a 28-year-old data scientist.");

PersonInfo personInfo = (await updates.ToAgentResponseAsync()).Deserialize<PersonInfo>(JsonSerializerOptions.Web);
Console.WriteLine($"Name: {personInfo.Name}, Age: {personInfo.Age}, Occupation: {personInfo.Occupation}");
// </structured_output_streaming>

await aiProjectClient.Agents.DeleteAgentAsync(agent.Name);

namespace SampleApp
{
    [Description("Information about a person including their name, age, and occupation")]
    public class PersonInfo
    {
        [JsonPropertyName("name")]
        public string? Name { get; set; }

        [JsonPropertyName("age")]
        public int? Age { get; set; }

        [JsonPropertyName("occupation")]
        public string? Occupation { get; set; }
    }
}
